engine/src/main/java/org/pentaho/di/repository/RepositoriesMeta.java

```diff
@@ -0,0 +1,409 @@
+      }
+      String id = XMLHandler.getTagValue( repnode, "id" );
+      if ( Utils.isEmpty( id ) ) {
+        // Backward compatibility : if the id is not defined, it's the database repository!
+        //
+        id = KettleDatabaseRepositoryMeta.REPOSITORY_TYPE_ID;
+      }
+      try {
+        RepositoryMeta repositoryMeta =
+          PluginRegistry.getInstance().loadClass( RepositoryPluginType.class, id, RepositoryMeta.class );
+        if ( repositoryMeta != null ) {
+          repositoryMeta.loadXML( repnode, databases );
+          // Backward compatibility. Description is now required as it will be what gets displayed on the
+          // repositories dialog
+          if ( repositoryMeta.getDescription() == null || repositoryMeta.getDescription().equals( "" ) ) {
+            repositoryMeta.setDescription( repositoryMeta.getName() );
+          }
+          addRepository( repositoryMeta );
+          if ( log.isDebug() ) {
+            log.logDebug( BaseMessages
+              .getString( PKG, "RepositoryMeta.Log.ReadRepository", repositoryMeta.getName() ) );
+          }
+        } else {
+          unableToReadIds.append( id );
+          unableToReadIds.append( "," );
+          if ( log.isDebug() ) {
+            log.logDebug( BaseMessages.getString( PKG, "RepositoryMeta.Error.ReadRepositoryId", id ) );
+          }
+        }
+      } catch ( KettleException ex ) {
+        // Get to the root cause
+        Throwable cause = ex;
+        kettleException = ex;
+        while ( cause.getCause() != null ) {
+          cause = cause.getCause();
+        }
+
+        if ( cause instanceof KettleRepositoryNotSupportedException ) {
+          // If the root cause is a KettleRepositoryNotSupportedException, do not fail
+          if ( log.isDebug() ) {
+            log.logDebug( BaseMessages.getString( PKG, "RepositoryMeta.Error.UnrecognizedRepositoryType", id ) );
+          }
+        }
+      }
+    }
+    if ( unableToReadIds != null && unableToReadIds.length() > 0 ) {
+      errorMessage =
+        BaseMessages.getString( PKG, "RepositoryMeta.Error.ReadRepositoryIdNotAvailable", unableToReadIds
+          .substring( 0, unableToReadIds.lastIndexOf( "," ) ) );
+    }
+    if ( kettleException != null ) {
+      throw kettleException;
+    }
+  }
+
+  public String getXML() {
+    String retval = "";
+
+    retval += "<?xml version=\"1.0\" encoding=\"UTF-8\"?>" + Const.CR;
+    retval += "<repositories>" + Const.CR;
+
+    for ( int i = 0; i < nrDatabases(); i++ ) {
+      DatabaseMeta conn = getDatabase( i );
+      retval += conn.getXML();
+    }
+
+    for ( int i = 0; i < nrRepositories(); i++ ) {
+      RepositoryMeta ri = getRepository( i );
+      retval += ri.getXML();
+    }
+
+    retval += "  </repositories>" + Const.CR;
+    return retval;
+  }
+
+  public void writeData() throws KettleException {
+    try {
+      FileOutputStream fos = new FileOutputStream( new File( getKettleUserRepositoriesFile() ) );
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *fos*. The resource is referenced at line 379. The resource is closed at line 380. There are other execution paths that do not contain closure statements, for example, when FileOutputStream.write throws an exception. Ensure *fos* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).